compiler/incremental-compilation-impl/src/org/jetbrains/kotlin/incremental/classpathDiff/ProgramSymbol.kt

```diff
@@ -20,97 +22,172 @@ data class ClassMember(val classId: ClassId, val memberName: String) : ProgramSy
         }
+        is ClassMember -> LookupSymbol(name = memberName, scope = classId.asSingleFqName().asString())
+        is PackageMember -> LookupSymbol(name = memberName, scope = packageFqName.asString())
     }
-    return PackageMemberSet(combinedMap)
 }
 
 /**
- * Finds [LookupSymbol]s that potentially refer to classes on the given classpath, and returns the [ProgramSymbol]s corresponding to those
- * [LookupSymbol]s.
+ * Converts [LookupSymbol]s to [ProgramSymbol]s.
+ *
+ * Since [LookupSymbol]s are ambiguous, we need to use the given classes to disambiguate them.
  *
- * Note: Some [LookupSymbol]s may refer to a class outside the classpath (e.g., `java/lang/Object`, or a class in the sources being
- * compiled). The returned result will not include those symbols.
+ * A [LookupSymbol] may be converted to more than one [ProgramSymbol]. For example, given this class:
+ *    class Foo {
+ *        class Bar
+ *        fun Bar(x: Int) {}
+ *     }
+ * LookupSymbol(scope = "Foo", name = "Bar") will be converted to both ClassSymbol( "Foo.Bar") and ClassMember("Foo", "Bar").
  *
- * The given classpath must not contain duplicate classes.
+ * If a [LookupSymbol] does not refer to any symbols in the given classes, it will be ignored.
  *
- * It's okay to over-approximate the result.
+ * Note: It's okay to over-approximate the result.
  */
-internal fun Collection<LookupSymbol>.filterLookupSymbols(classpath: List<AccessibleClassSnapshot>): List<ProgramSymbol> {
-    val regularClassesOnClasspath: List<ClassId> = classpath.mapNotNull {
-        when (it) {
-            is RegularKotlinClassSnapshot -> it.classId
-            is JavaClassSnapshot -> it.classId
-            is PackageFacadeKotlinClassSnapshot, is MultifileClassKotlinClassSnapshot -> null
+internal fun Collection<LookupSymbol>.toProgramSymbolSet(allClasses: Iterable<AccessibleClassSnapshot>): ProgramSymbolSet {
+    // Use LookupSymbolSet for efficiency
+    val lookupSymbols = LookupSymbolSet(this)
+
+    val collector = ProgramSymbolSet.Collector()
+    allClasses.forEach { clazz ->
+        when (clazz) {
+            is RegularKotlinClassSnapshot, is JavaClassSnapshot -> {
+                // Collect ClassSymbols
+                if (ClassSymbol(clazz.classId).toLookupSymbol() in lookupSymbols) {
+                    collector.addClass(clazz.classId)
+                }
+
+                // Collect ClassMembers
+                // To do this correctly, we would need to get the intersection of clazz.classMembers and lookupSymbolsInScope (it's possible
+                // that two different kinds of ProgramSymbols have the same LookupSymbol/scope -- see the example in this method's kdoc).
+                // However, we don't have information about clazz.classMembers, so we'll take all of lookupSymbolsInScope (it's okay to
+                // over-approximate the result).
+                val lookupSymbolsInScope = lookupSymbols.getMembersInScope(clazz.classId.asSingleFqName())
```

Comment:
What happens if lookupSymbolsInScope is empty?

Reply:
It will be a no-op inside `ProgramSymbolSet.Collector#addClassMembers`.